### ISSUE TITLE
Support `torch.linalg.einsum`

### ISSUE THREAD

**Issue by AnirudhDagar** [pytorch/pytorch#62812]
## 🚀 Feature
Support `torch.linalg.einsum`.

cc @mruberry @rgommers

## Motivation & Pitch

Python Array API proposes the use of `linalg.einsum` among other linear algebra functions. Currently, PyTorch supports the same functionality with `torch.einsum`. Array-API Ops tracker (#58742) lists a few aliases and it would be nice to add an alias `torch.linalg.einsum` to `torch.einsum` making it compliant with Array API.

## Discussion
Array-API still has a [TODO](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#einsum) for `einsum`. However, I guess we can still add the alias because it will still go in the `linalg` namespace. Maybe @rgommers can comment here before I go ahead. Thanks!

## Additional context

* [Python Array API linalg.einsum](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html#linalg-einsum)

**Comment by rgommers** [comment]
> Array-API still has a [TODO](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#einsum) for `einsum`. However, I guess we can still add the alias because it will still go in the `linalg` namespace.

Yes we should hurry up and add that. I would suggest leaving the `einsum` alias till the end, just to avoid potential issues.

**Comment by rgommers** [comment]
Update: we actually left `einsum` out of the 2021 standard, because it turned out to be pretty complex to specify in detail.

That doesn't mean it is not a good idea to add `einsum` into `torch.linalg` at some point, but it's not a priority I'd say. So I'll go ahead and remove the array API label.

**Comment by yaroslavvb** [comment]
@rgommers do you have a link to the discussion about einsum complexity?

Old-fashioned "einsum" (as-in, what Einstein used) restricted each index to occur in 1 or 2 tensors.
Sometimes people further assume that einsum index will occur in exactly 2 tensors, opt_einsum has some [unexpected behavior](https://github.com/dgasmith/opt_einsum/issues/112) due to this assumption.
A general einsum is a bipartite graph between "index nodes" and "tensor nodes"

Currently np.einsum implements something halfway between "old-fashioned" and "general"

**Comment by rgommers** [comment]
> @rgommers do you have a link to the discussion about einsum complexity?

I can't find a discussion on an issue tracker, but it's easy to see when you look at [the numpy.einsum docs](https://numpy.org/devdocs/reference/generated/numpy.einsum.html#numpy.einsum) or the [opt_einsum package docs](https://optimized-einsum.readthedocs.io/en/stable/) that there's a large number of options and four different algorithms which can be chosen via the `optimize=` keyword. Deciding which algorithms to include would be nontrivial, and if we added all four then the function specification would be by far the largest one in the whole array API standard specification.

**Comment by yaroslavvb** [comment]
I see, that's a good point. There are a lot more potential methods implemented in https://github.com/jcmgray/cotengra
Deciding on an `optimize` option in `einsum` seems similar to deciding on `lapack_driver` option in `linalg.svd`

IE
`linalg.svd(...,lapack_driver='gesvd'")`

The difference is that "gesvd" algorithm has been around since Lapack days, meanwhile these einsum optimizer algorithms are new

**Comment by leofang** [comment]
Please also check out cuTensorNet, part of the cuQuantum SDK. We provide Python bindings that accept NumPy, CuPy and PyTorch tensors:
https://docs.nvidia.com/cuda/cuquantum/python/api/cuquantum.html#python-functions
I have been thinking maybe it's best for `einsum` to implement open slots (ex: the optimize keyword) for venders to provide pluggable algorithms. But then it would not be portable... However, by limiting the available algorithms the experience from CoTenGra and cuTensorNet have shown that there could be huge perf diff in terms of both the quality of the contraction path and how it is actually executed. I do not have good answers for now.

**Comment by yaroslavvb** [comment]
I think the broader issue with einsum "optimize='optimal'" kwarg is that can't ask for an "optimal" setting without also specifying the class of problems for which it should work.

For instance, existing np.einsum "optimal" can't solve the Matrix Chain problem with 15 matrices in reasonable time ([colab](https://colab.research.google.com/drive/1W4QB09Yd6-otJTS9Lk7qj1wTQVy9MAJ2#scrollTo=qP2Ubjeaxsmr)), so if you add this algorithm to the API, you are essentially specifying a "default" class of problems for which it works. This default class excludes matrix chain problems of size 15. But why? There's a polynomial time algorithm to solve it [for chains]((https://en.wikipedia.org/wiki/Matrix_chain_multiplication#A_dynamic_programming_algorithm)), and more generally [for trees](https://pubmed.ncbi.nlm.nih.gov/31770963/).

Another example, `opt_einsum(..method=dp)' can exactly optimize some large tree-structured problems which np.einsum can't, but it requires specific array dimensions ([example](https://colab.research.google.com/drive/1W4QB09Yd6-otJTS9Lk7qj1wTQVy9MAJ2#scrollTo=pD9M8Tx6zCqG))

More generally, there are exact tractable algorithms to perform this kind of optimization for special structures of tensor networks. For instance, bounded treewidth graphs, planar graphs and more generally minor-excluded graphs. It seems unlikely we'll have a single "optimize='optimal'" algorithm for all einsum types. 

@leofang how good is cuTensorNet support for exact contraction optimization? For instance, can it optimize tree-structured einsums? Here's an [example](https://github.com/dgasmith/opt_einsum/issues/112#issuecomment-559957281) that opt_einsum had problem with.

Exact contraction optimization is an important use-case for non-physicists. For instance, neural network computations tend to involve a lot of small einsums that are optimizable exactly. I looked into using 'opt_einsum' for this, but ran into a few issues (for instance [here](https://github.com/dgasmith/opt_einsum/issues/112) and [here](https://github.com/dgasmith/opt_einsum/issues/167))

**Comment by leofang** [comment]
Thanks, @yaroslavvb, for your valuable inputs. We've started investigating using cuTensorNet on the TN examples that you provided. It's good for us to understand better potential use cases. To not pollute this issue tracker, however, I'd suggest us to exchange offline via email (cuquantum-python@nvidia.com). We'll be more than happy to follow up with you 🙂

Brief comments relevant in terms of API design/implementation: 
> I think the broader issue with einsum "optimize='optimal'" kwarg is that can't ask for an "optimal" algorithm without also specifying the class of problems for which it should work.

To me it's just a matter of naming. It is a working algorithm no doubt, but may not be efficient for all TN classes. I'd just call it "algorithm X" (or whatever you like) and offer it as one of the available algorithms in `einsum()` if I were implementing it in PyTorch or whatever library. But, to me what you wish for is a sort of TN preprocessor to determine the best algorithm. I feel this should be an opt-in feature, not default behavior.

One minor question on the  Matrix Chain problem: Are you sure your einsum string was constructed correctly? Shouldn't it be
```
ij,jk,kl,lm,mn,no,op,pq,qr->ir  # notice the last mode
```
instead of
```
ij,jk,kl,lm,mn,no,op,pq,qr->iq
```
?

> how good is cuTensorNet support for exact contraction optimization? For instance, can it optimize tree-structured einsums? 

We currently only support exact contraction and no approximate TNs yet. So I'd say yes we can handle it, but as I said we're looking into your examples. Let's move this part of discussion offline 🙂 Thanks.
